File: internal/unit/evidence.go
package unit

import (
	"regexp"
	"sort"
	"strconv"
	"strings"
	"unicode/utf16"

	"github.com/clipperhouse/uax29/v2/sentences"
	s "github.com/mardwerk/unit-generator/internal/schema"
)

// EvidenceSpan is a quotable source passage with a stable ID.
type EvidenceSpan struct {
	ID         string `json:"id"`
	DocumentID string `json:"documentId"`
	Text       string `json:"text"`
}

// TechniqueContext is the period context of a retrieved technique article.
type TechniqueContext struct {
	DocumentID string `json:"documentId"`
	Quote      string `json:"quote"`
	Technique  string `json:"technique"`
}

var formerWord = regexp.MustCompile(`(?i)\bformer\b`)

// jsLines splits text at JavaScript line terminators, like a multiline regex.
func jsLines(text string) []string {
	return strings.FieldsFunc(text, func(r rune) bool {
		return r == '\n' || r == '\r' || r == 0x2028 || r == 0x2029
	})
}

// firstLine returns the first line starting with prefix and having content after it.
func firstLine(text, prefix string) (string, bool) {
	for _, line := range jsLines(text) {
		if strings.HasPrefix(line, prefix) && len(line) > len(prefix) {
			return line, true
		}
	}
	return "", false
}

// HistoricalTechniqueContext keeps the explicit period of an observed technique link.
func HistoricalTechniqueContext(request *Request, documentID string) *TechniqueContext {
	var document *Document
	for i := range request.Documents {
		if request.Documents[i].ID == documentID {
			document = &request.Documents[i]
			break
		}
	}
	if document == nil || !strings.HasPrefix(document.ID, "character-technique:") {
		return nil
	}
	section, ok := firstLine(document.Text, "Section: ")
	if !ok || !formerWord.MatchString(section) {
		return nil
	}
	technique := "Selected technique"
	if line, ok := firstLine(document.Text, "Link text: "); ok {
		technique = strings.TrimPrefix(line, "Link text: ")
	}
	return &TechniqueContext{DocumentID: documentID, Quote: section, Technique: technique}
}

// u16 is a JavaScript string: UTF-16 code units.
type u16 []uint16

func toU16(text string) u16  { return utf16.Encode([]rune(text)) }
func (t u16) String() string { return string(utf16.Decode(t)) }
func (t u16) trim() u16      { return toU16(s.Trim(t.String())) }
func (t u16) lastSpace(from int) int {
	if from >= len(t) {
		from = len(t) - 1
	}
	for i := from; i >= 0; i-- {
		if t[i] == ' ' {
			return i
		}
	}
	return -1
}

// EvidenceSpans splits source documents into deterministic, quotable passages.
func EvidenceSpans(request *Request) []EvidenceSpan {
	var spans []EvidenceSpan
	for docIndex, document := range request.Documents {
		if document.Kind != "source" {
			continue
		}
		var passages []string
		segments := sentences.FromString(document.Text)
		for segments.Next() {
			segment := segments.Value()
			if n := len(passages); n > 0 && s.UTF16Len(s.Trim(passages[n-1])) < 15 {
				passages[n-1] += segment
			} else {
				passages = append(passages, segment)
			}
		}
		if n := len(passages); n > 1 && s.UTF16Len(s.Trim(passages[n-1])) < 15 {
			tail := passages[n-1]
			passages = passages[:n-1]
			passages[n-2] += tail
		}
		index := 0
		for _, passage := range passages {
			remaining := toU16(passage).trim()
			for len(remaining) > 0 {
				end := len(remaining)
				if end > 450 {
					boundary := remaining.lastSpace(450)
					if boundary >= 15 {
						end = boundary
					} else {
						end = 450
					}
					if s.UTF16Len(s.Trim(remaining[end:].String())) < 15 {
						end = max(15, end-15)
					}
				}
				text := remaining[:end].trim()
				if len(text) > 0 {
					spans = append(spans, EvidenceSpan{
						ID:         "source" + strconv.Itoa(docIndex+1) + ":" + strconv.Itoa(index),
						DocumentID: document.ID,
						Text:       text.String(),
					})
					index++
				}
				remaining = remaining[end:].trim()
			}
		}
	}
	return spans
}

var (
	combatWords   = regexp.MustCompile(`(?i)\b(?:abilit\w*|power\w*|attack\w*|combat|strength|speed|technique\w*|transform\w*|form\w*|damage|control|weapon\w*|punch\w*|beam\w*|stretc\w*|absor\w*|mimic\w*|summon\w*|limit\w*|weak\w*|cannot|unable|immune|immunity)\b`)
	signatureWord = regexp.MustCompile(`(?i)\b(?:signature|primary|characteristic)\b`)
	actionWords   = regexp.MustCompile(`(?i)\b(?:fires?|firing|shoots?|shooting|launch\w*|emits?|emitting|strikes?|striking|throws?|throwing|projectiles?|slashes?|slashing)\b`)
	limitWords    = regexp.MustCompile(`(?i)\b(?:cannot|unable|requires?|only|former|limitations?|weakness\w*)\b`)
)

// AuthorEvidence is a bounded lexical selection of spans for the model.
func AuthorEvidence(request *Request) []EvidenceSpan {
	all := EvidenceSpans(request)
	const limit, maxSpans = 18000, 96
	total := 0
	for _, span := range all {
		total += s.UTF16Len(span.Text)
	}
	if len(all) <= maxSpans && total <= limit {
		return all
	}
	type ranked struct {
		span  EvidenceSpan
		index int
		score int
	}
	first := map[string]bool{}
	seen := map[string]bool{}
	list := make([]ranked, len(all))
	for i, span := range all {
		identity := !first[span.DocumentID]
		first[span.DocumentID] = true
		key := span.DocumentID + "\x00" + span.Text
		repeated := seen[key]
		seen[key] = true
		combat := len(combatWords.FindAllStringIndex(span.Text, -1))
		relevance := 0
		if signatureWord.MatchString(span.Text) {
			relevance += 3
		}
		relevance += min(len(actionWords.FindAllStringIndex(span.Text, -1)), 3)
		if limitWords.MatchString(span.Text) {
			relevance += 4
		}
		score := -1
		if !repeated {
			score = min(combat, 12) + relevance
			if identity {
				score += 100
			}
		}
		list[i] = ranked{span, i, score}
	}
	sort.SliceStable(list, func(a, b int) bool {
		if list[a].score != list[b].score {
			return list[a].score > list[b].score
		}
		return list[a].index < list[b].index
	})
	used, selected := 0, 0
	reserved := map[string]bool{}
	for _, document := range request.Documents {
		if !strings.HasPrefix(document.ID, "character-technique:") {
			continue
		}
		var packet []EvidenceSpan
		size := 0
		for _, span := range all {
			if span.DocumentID == document.ID {
				packet = append(packet, span)
				size += s.UTF16Len(span.Text)
			}
		}
		if len(packet) > 32 || size > 6000 || selected+len(packet) > maxSpans || used+size > limit {
			continue
		}
		for _, span := range packet {
			reserved[span.ID] = true
		}
		used += size
		selected += len(packet)
	}
	var kept []ranked
	for _, item := range list {
		if reserved[item.span.ID] {
			kept = append(kept, item)
			continue
		}
		length := s.UTF16Len(item.span.Text)
		if selected >= maxSpans || used+length > limit {
			continue
		}
		used += length
		selected++
		kept = append(kept, item)
	}
	sort.SliceStable(kept, func(a, b int) bool { return kept[a].index < kept[b].index })
	out := make([]EvidenceSpan, len(kept))
	for i, item := range kept {
		out[i] = item.span
	}
	return out
}

File: src/core/planned-v1/evidence.ts
import type { AuthorRequest } from '../schemas.js';

export interface EvidenceSpan {
  id: string;
  documentId: string;
  text: string;
}

/** Preserve the explicit period context carried by an observed technique link. */
export function historicalTechniqueContext(request: AuthorRequest, documentId: string) {
  const document = request.documents.find((entry) => entry.id === documentId);
  if (!document?.id.startsWith('character-technique:')) return null;
  const section = document.text.match(/^Section: (.+)$/m)?.[0];
  if (!section || !/\bformer\b/i.test(section)) return null;
  const technique = document.text.match(/^Link text: (.+)$/m)?.[1] ?? 'Selected technique';
  return { documentId, quote: section, technique };
}

/** Deterministic source passages. Selection is model-assisted; quotation is not. */
export function evidenceSpans(request: AuthorRequest): EvidenceSpan[] {
  const spans: EvidenceSpan[] = [];
  const segmenter = new Intl.Segmenter('en', { granularity: 'sentence' });
  request.documents.forEach((document, docIndex) => {
    if (document.kind !== 'source') return;
    const passages: string[] = [];
    for (const { segment } of segmenter.segment(document.text)) {
      const previous = passages.at(-1);
      if (previous !== undefined && previous.trim().length < 15) {
        passages[passages.length - 1] = previous + segment;
      } else {
        passages.push(segment);
      }
    }
    if (passages.length > 1 && passages.at(-1)!.trim().length < 15) {
      const tail = passages.pop()!;
      passages[passages.length - 1] += tail;
    }
    let index = 0;
    for (const passage of passages) {
      let remaining = passage.trim();
      while (remaining.length) {
        let end = remaining.length;
        if (end > 450) {
          const wordBoundary = remaining.lastIndexOf(' ', 450);
          end = wordBoundary >= 15 ? wordBoundary : 450;
          // Leave enough of the preceding text with a short final clause or word.
          if (remaining.slice(end).trim().length < 15) end = Math.max(15, end - 15);
        }
        const text = remaining.slice(0, end).trim();
        // Even a wholly short document remains visible to the author. It cannot
        // meet a longer quotation requirement merely by being omitted here.
        if (text)
          spans.push({ id: `source${docIndex + 1}:${index++}`, documentId: document.id, text });
        remaining = remaining.slice(end).trim();
      }
    }
  });
  return spans;
}

/** Bounded lexical retrieval, not a guarantee of canon relevance or source coverage.
 * Full documents remain in the Request for inspection. */
export function authorEvidence(request: AuthorRequest): EvidenceSpan[] {
  const all = evidenceSpans(request);
  const planned = request.mechanicsDefinition?.profile.authoringMode === 'planned-v1';
  const limit = planned ? 18_000 : 6_000;
  const maxSpans = planned ? 96 : Infinity;
  if (all.length <= maxSpans && all.reduce((total, span) => total + span.text.length, 0) <= limit)
    return all;
  const first = new Set<string>();
  const seenText = new Set<string>();
  const ranked = all.map((span, index) => {
    const identity = !first.has(span.documentId);
    first.add(span.documentId);
    const textKey = JSON.stringify([span.documentId, span.text]);
    const repeated = seenText.has(textKey);
    seenText.add(textKey);
    const combat =
      span.text.match(
        /\b(?:abilit\w*|power\w*|attack\w*|combat|strength|speed|technique\w*|transform\w*|form\w*|damage|control|weapon\w*|punch\w*|beam\w*|stretc\w*|absor\w*|mimic\w*|summon\w*|limit\w*|weak\w*|cannot|unable|immune|immunity)\b/gi,
      )?.length ?? 0;
    const planningRelevance = planned
      ? (/\b(?:signature|primary|characteristic)\b/i.test(span.text) ? 3 : 0) +
        Math.min(
          span.text.match(
            /\b(?:fires?|firing|shoots?|shooting|launch\w*|emits?|emitting|strikes?|striking|throws?|throwing|projectiles?|slashes?|slashing)\b/gi,
          )?.length ?? 0,
          3,
        ) +
        (/\b(?:cannot|unable|requires?|only|former|limitations?|weakness\w*)\b/i.test(span.text)
          ? 4
          : 0)
      : 0;
    return {
      span,
      index,
      score:
        planned && repeated ? -1 : (identity ? 100 : 0) + Math.min(combat, 12) + planningRelevance,
    };
  });
  ranked.sort((a, b) => b.score - a.score || a.index - b.index);
  let used = 0;
  let selectedCount = 0;
  const reserved = new Set<string>();
  if (planned) {
    // Retrieved technique articles already pair ownership and period context with
    // behavior and exceptions. Keep compact articles whole before lexical ranking
    // can favor repeated inventory headings over their less formulaic prose.
    for (const document of request.documents) {
      if (!document.id.startsWith('character-technique:')) continue;
      const packet = all.filter((span) => span.documentId === document.id);
      const size = packet.reduce((total, span) => total + span.text.length, 0);
      if (
        packet.length > 32 ||
        size > 6_000 ||
        selectedCount + packet.length > maxSpans ||
        used + size > limit
      )
        continue;
      for (const span of packet) reserved.add(span.id);
      used += size;
      selectedCount += packet.length;
    }
  }
  const selected = ranked.filter(({ span }) => {
    if (reserved.has(span.id)) return true;
    if (selectedCount >= maxSpans || used + span.text.length > limit) return false;
    used += span.text.length;
    selectedCount++;
    return true;
  });
  return selected.sort((a, b) => a.index - b.index).map(({ span }) => span);
}
